### PR TITLE
import, refactor, improve & yarn.lock support

### PR DESCRIPTION
**Pull request by obfusk**

### CODE REVIEW

File: mklock.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+const lockfile      = require("@yarnpkg/lockfile")
+const semver        = require("semver")
+const ssri          = require("ssri")
+
+// Usage: node mklock.js package-lock.json package.json yarn.lock integrities.json
+
+let pkgLockFile     = process.argv[2]
+let pkgFile         = process.argv[3]
+let yarnLockFile    = process.argv[4]
+let intFile         = process.argv[5]
```
Comment (yorickvP) [inline mklock.js]: `let [_node, _file, pkgLockFile, pkgFile, yarnLockFile, intFile] = process.argv`

File: mklock.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  if (key[0] == "@") {
+    let [name, vsn] = key.slice(1).split("@")
+    return ["@"+name, vsn]
+  } else {
+    return key.split("@")
+  }
+}
+
+function addDeps(obj, pkg, vsn, seen) {
+  let pkgvsn            = pkg + "@" + vsn
```
Comment (yorickvP) [inline mklock.js]: how about more consts?

File: default.nix [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        if [ "$useYarnLock" = "1" ]; then
+          cp ${lockfile} ${builtins.baseNameOf pkgLockJson}
+          chmod u+w ${builtins.baseNameOf pkgLockJson}
+        fi
+        addToSearchPath NODE_PATH ${npmModules}
+        node ${./mkcache.js} ${npmCacheInput lock}
+      '';
+
+      buildPhase = ''
+        set -e
```
Comment (yorickvP) [inline default.nix]: These are called as functions in an existing bash script with `set -euo pipefail`:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh

File: default.nix [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+      lock            = fromJSON (readFile lockfile);
+
+      name            = "${info.name}-${info.version}";
+      npm             = "${nodejs-10_x}/bin/npm";
+      npmAlias        = ''npm() { ${npm} "$@" $npmFlags; }'';
+      npmModules      = "${nodejs-10_x}/lib/node_modules/npm/node_modules";
+
+      mkLockFileFromYarn = runCommand "yarn-package-lock.json" {} ''
+        set -e
+        addToSearchPath NODE_PATH ${npmModules}
```
Comment (yorickvP) [inline default.nix]: nice! I didn't know about this

File: default.nix [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+      yarnIntFile     = writeText "integreties.json" (toJSON yarnIntegreties);
+      lockfile        = if useYarnLock then mkLockFileFromYarn else pkgLockJson;
+
+      info            = fromJSON (readFile pkgJson);
+      lock            = fromJSON (readFile lockfile);
+
+      name            = "${info.name}-${info.version}";
+      npm             = "${nodejs-10_x}/bin/npm";
+      npmAlias        = ''npm() { ${npm} "$@" $npmFlags; }'';
```
Comment (yorickvP) [inline default.nix]: it's possible to pass npm arguments using env variables, but I didn't want to break nix-shell. Given sufficient divergence between shell and build, maybe they should be split.

File: default.nix (hunk 3 above)
Comment (obfusk) [reply default.nix]: Good to know. Wanted to check on that. Thanks.

File: default.nix (hunk 5 above)
Comment (obfusk) [reply default.nix]: Could you elaborate on that?

File: mklock.js (hunk 2 above)
Comment (obfusk) [reply mklock.js]: Yeah. Thanks for the reminder. `let` feels more natural if you're used to lisp, but here `const` is indeed better.

File: mklock.js (hunk 1 above)
Comment (obfusk) [reply mklock.js]: `let [pkgLockFile, pkgFile, yarnLockFile, intFile] = process.argv.slice(2)`

**Pull request closed without merging**